test/test_train_mp_mnist.py

```diff
@@ -0,0 +1,160 @@
+            [transforms.ToTensor(),
+             transforms.Normalize((0.1307,), (0.3081,))]))
+    test_dataset = datasets.MNIST(
+        FLAGS.datadir,
+        train=False,
+        transform=transforms.Compose(
+            [transforms.ToTensor(),
+             transforms.Normalize((0.1307,), (0.3081,))]))
+    train_sampler = None
+    if xm.xrt_world_size() > 1:
+      train_sampler = torch.utils.data.distributed.DistributedSampler(
+          train_dataset,
+          num_replicas=xm.xrt_world_size(),
+          rank=xm.get_ordinal(),
+          shuffle=True)
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset,
+        batch_size=FLAGS.batch_size,
+        sampler=train_sampler,
+        shuffle=False if train_sampler else True,
+        num_workers=FLAGS.num_workers)
+    test_loader = torch.utils.data.DataLoader(
+        test_dataset,
+        batch_size=FLAGS.batch_size,
+        shuffle=False,
+        num_workers=FLAGS.num_workers)
+
+  # Scale learning rate to num cores
+  lr = FLAGS.lr * xm.xrt_world_size()
+
+  device = xm.xla_device()
+  model = MNIST().to(device)
+  optimizer = optim.SGD(model.parameters(), lr=lr, momentum=FLAGS.momentum)
+  loss_fn = nn.NLLLoss()
+
+  def train_loop_fn(loader):
+    tracker = xm.RateTracker()
+
+    model.train()
+    for x, (data, target) in loader:
+      optimizer.zero_grad()
+      output = model(data)
+      loss = loss_fn(output, target)
+      loss.backward()
+      xm.optimizer_step(optimizer)
+      tracker.add(FLAGS.batch_size)
+      if x % FLAGS.log_steps == 0:
+        test_utils.print_training_update(device, x, loss.item(), tracker.rate(),
+                                         tracker.global_rate())
+
+  def test_loop_fn(loader):
+    total_samples = 0
+    correct = 0
+    model.eval()
+    for x, (data, target) in loader:
+      output = model(data)
+      pred = output.max(1, keepdim=True)[1]
+      correct += pred.eq(target.view_as(pred)).sum().item()
+      total_samples += data.size()[0]
+
+    accuracy = 100.0 * correct / total_samples
+    test_utils.print_test_update(device, accuracy)
+    return accuracy
+
+  accuracy = 0.0
+  for epoch in range(1, FLAGS.num_epochs + 1):
+    para_loader = dp.ParallelLoader(train_loader, [device])
+    train_loop_fn(para_loader.per_device_loader(device))
+
+    para_loader = dp.ParallelLoader(test_loader, [device])
+    accuracy = test_loop_fn(para_loader.per_device_loader(device))
+    if FLAGS.metrics_debug:
+      print(torch_xla._XLAC._xla_metrics_report())
+
+  return accuracy
+
+
+def _mp_fn(index, flags):
+  global FLAGS
+  FLAGS = flags
+  torch.set_default_tensor_type('torch.FloatTensor')
+  accuracy = train_mnist()
+  if FLAGS.tidy and os.path.isdir(FLAGS.datadir):
+    shutil.rmtree(FLAGS.datadir)
+  if accuracy < FLAGS.target_accuracy:
+    print('Accuracy {} is below target {}'.format(accuracy,
+                                                  FLAGS.target_accuracy))
```

Comment:
Just so I understand:
the code in _mp_fn would run only once during setup, e.g. a user would see this 'Accuracy {} is below target...' print statement only once using --num_cores=8, but the code in train_mnist() would run once per device?

I guess to me it's not clear at which point in this file the fork occurs

Reply:
The mp_fn runs per device. Each process has its own device.
If we want to print only from one device, we can use ordinal==0 checks.

Reply:
Does that mean that each device would have a different accuracy? That would mean some devices could pass the test and some could fail.

Or do we sync the model weights before computing accuracy, and each device runs over the entire test set, so each device should have the same accuracy?

Reply:
Since we are running the whole validation set for each device, they should be getting same accuracy or very close.